Refactor `EditPostActivity` Authentication and Enhance Cookie Authenticator

oguzkocer opened the pull request:
## Summary

This PR refactors the authentication logic in `EditPostActivity` to follow MVVM patterns and enhances the `WordPressCookieAuthenticator` that was introduced in the previous PR. The main changes include moving the authenticator to the WordPress module, adding comprehensive caching, and extracting authentication logic to a ViewModel.

### Key Changes

- **Extract authentication logic to ViewModel**: Move authentication logic from `EditPostActivity` to `EditPostAuthViewModel` using sealed classes for clean state management
- **Move and enhance WordPressCookieAuthenticator**: Relocate from `libs/editor` to WordPress module (`org.wordpress.android.auth`) with significant improvements:
  - Remove Java callback support (now Kotlin-only with coroutines)
  - Add 6-month cookie caching with SharedPreferences to avoid repeated authentication
  - Add @Singleton annotation and improved dependency injection with named parameters
  - Add cache management methods for logout cleanup
  - Change logging from `T.EDITOR` to `T.API` for better categorization
  - Extract request execution logic into separate validated methods
- **Add logout cleanup**: Clear cached cookies on logout in `AppInitializer`
- **Add comprehensive unit tests**: Full test coverage for both ViewModel and authenticator with proper HTTP mocking

### Technical Details

- Uses `EditPostAuthViewModel` with sealed class states (`Idle`, `Success`, `Error`)
- Cookie caching implemented similar to `PrivateAtomicCookie` pattern with 6-month expiration
- Adds `ValidatedAuthParams` for type-safe parameter passing after validation
- All tests properly mock HTTP requests (no real network calls)
- Maintains authentication flow compatibility while simplifying the API (Kotlin-only)

## Testing Instructions

### Prerequisites
- Build and install the app on a test device
- Have access to both private atomic and private simple WordPress.com sites
- Test with different user accounts to verify caching behavior

### Test Scenarios

#### 1. Private Atomic WordPress.com Sites
**Default Editor:**
1. Open a post from a private atomic site in the default editor
2. Verify the editor loads successfully without authentication errors
3. Check that media content (images, videos) loads properly in the editor
4. Edit and save the post to ensure functionality works end-to-end

**Experimental Editor (if available):**
1. Switch to experimental editor for the same private atomic site
2. Open the same or different post
3. Verify content loads correctly including media
4. Test editing and saving functionality

#### 2. Private Simple WordPress.com Sites  
**Default Editor:**
1. Open a post from a private simple WordPress.com site in the default editor
2. Verify the editor loads without authentication issues
3. Confirm all media content displays correctly
4. Test post editing and saving

**Experimental Editor (if available):**
1. Use experimental editor with the private simple site
2. Open various posts and verify content loading
3. Test media handling and post operations

#### 3. Caching Behavior Verification
1. Open a post from a private site (triggers authentication)
2. Close and reopen the editor within 6 months
3. Verify faster loading (should use cached cookies)
4. Log out and log back in
5. Confirm authentication is required again (cache cleared on logout)

#### 4. Error Handling
1. Test with invalid credentials or network issues
2. Verify appropriate error messages are displayed
3. Confirm the app gracefully handles authentication failures

### Expected Results
- All content loads properly in both editors for private sites
- Media content displays correctly without authentication issues
- Caching improves performance on subsequent editor opens
- No regressions in existing authentication flows
- Error states are handled gracefully

### Notes
- Monitor app logs for authentication-related messages during testing
- Pay particular attention to media loading as this was a primary use case
- Test different network conditions to verify caching behavior

oguzkocer committed:
refactor: Move WordPressCookieAuthenticator to WordPress module

Move WordPressCookieAuthenticator from libs:editor to WordPress/auth to enable dependency injection.

Changes:
- Move class from libs/editor to WordPress/auth package
- Update to use @Named dependency injection for OkHttpClient and CoroutineDispatcher
- Update EditPostActivity import
- Remove unused ApplicationModule providers
- Simplify tests to only include parameter validation

oguzkocer committed:
Remove no longer necessary libs/editor/build.gradle dependency additions

oguzkocer committed:
refactor: Extract authentication logic to ViewModel with sealed classes

Moved cookie authentication from Activity to dedicated ViewModel for better
architecture and state management.

Changes:
- Extract EditPostAuthViewModel with sealed class states
- Remove callback complexity from WordPressCookieAuthenticator
- Use coroutines and postValue() for thread safety
- Consolidate cookie logic with single source of truth
- Fix Idle state handling to prevent premature UI setup
- Use functional programming patterns for cookie processing

This improves maintainability, testability, and follows modern Android
architecture patterns with proper separation of concerns.

oguzkocer committed:
refactor: Simplify cookie logic and add comprehensive unit tests

Refactor getCookiesForPrivateSites method to use cleaner when expression
that prioritizes checking if site is private first, making the logic more
readable and maintainable.

Add comprehensive unit tests for EditPostAuthViewModel covering:
- Authentication state transitions (Loading, Success, Error)
- Cookie processing for different site types (WPCom, atomic, private/public)
- Edge cases including authentication failures and missing cookies
- Cookie filtering to only include wordpress_logged_in cookies

Changes:
- Simplify conditional logic in getCookiesForPrivateSites method
- Add EditPostAuthViewModelTest with 11 comprehensive test cases
- Use mock-based approach for PrivateAtomicCookie in tests

oguzkocer committed:
feat: Add caching and logout cleanup to WordPressCookieAuthenticator

- Implement cookie caching with 6-hour expiration using SharedPreferences
- Add cache invalidation on corrupted data and expired entries
- Include logout cleanup in AppInitializer to clear cached cookies
- Update logging to use API tag and specify WordPress.com account cookies
- Fix unit tests to properly mock HTTP requests instead of making real network calls
- Add comprehensive test coverage for caching scenarios, HTTP errors, and network failures

Changes:
- Add @Singleton annotation and PreferenceUtilsWrapper dependency
- Implement getCachedCookies(), cacheCookies(), clearCachedCookies(), and clearAllCachedCookies() methods
- Add cache-first logic in authenticateForCookies() with automatic caching of successful results
- Integrate clearAllCachedCookies() call in AppInitializer.removeWpComUserRelatedData()
- Replace real OkHttpClient with mocked version in tests to prevent actual network requests
- Add 3 new test cases covering successful authentication, HTTP errors, and network failures

oguzkocer committed:
refactor: Extract request execution to separate method in WordPressCookieAuthenticator

- Extract HTTP request execution logic from authenticateForCookies() into performAuthenticationRequest() helper method
- Add ValidatedAuthParams data class to represent validated parameters passed to helper method
- Improve separation of concerns: main method handles validation and caching orchestration, helper method handles HTTP execution
- Maintain existing functionality with clearer code organization

Changes:
- Add ValidatedAuthParams private data class for validated parameter passing
- Create performAuthenticationRequest() method containing HTTP request logic
- Update authenticateForCookies() to validate parameters then delegate to helper method
- All tests continue to pass, confirming refactoring preserves behavior

oguzkocer committed:
feat: Extend cookie cache expiration from 6 hours to 6 months

WordPress.com authentication cookies are now cached for 6 months instead of 6 hours to provide better user experience by reducing authentication requests.

dangermattic commented:
<!--
  0 Errors
  1 Warning: This PR is larger than 300 lin...
  1 Message: This PR is still a Draft: some...
  0 Markdowns
-->
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">This PR is larger than 300 lines of changes. Please consider splitting it into smaller PRs for easier and faster reviews.</td>
    </tr>
  </tbody>
</table>
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Message">
          1 Message
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:book:</td>
      <td data-sticky="false">This PR is still a Draft: some checks will be skipped.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/ui/posts/EditPostAuthViewModel.kt:
```diff
@@ -0,0 +1,97 @@
+                }
+
+                is WordPressCookieAuthenticator.AuthResult.Failure -> {
+                    AppLog.e(AppLog.T.EDITOR, "Failed to fetch cookies for Simple site: ${result.error}")
```
This is not correct as it's a different file and part of the editor flow.

oguzkocer commented on WordPress/src/main/java/org/wordpress/android/auth/WordPressCookieAuthenticator.kt:
```diff
@@ -0,0 +1,380 @@
+        private const val WP_LOGIN_URL = "https://wordpress.com/wp-login.php"
+        private const val CONTENT_TYPE_FORM = "application/x-www-form-urlencoded"
+        private const val WPCOM_COOKIE_CACHE_PREFIX = "WPCOM_COOKIE_CACHE_"
+        private const val COOKIE_CACHE_EXPIRATION_HOURS = 24 * 30 * 6 // 6 months (approximately)
```
That's exactly why it's **approximately** 6 months 😮‍💨

oguzkocer committed:
Fix flaky test that caused uncaught exceptions in CI

The test `fetchWpComCookies starts with Loading state` was causing
NoWhenBranchMatchedException to leak into other tests due to missing
mock setup for the authenticator.

The test was also fundamentally flawed because it tried to check the
Loading state immediately after calling fetchWpComCookies(), but
postValue() is asynchronous and the mocked authenticator would complete
immediately, potentially skipping the Loading state.

Changes:
- Renamed test to better reflect what it actually tests
- Added proper mock setup to prevent exception leakage
- Changed approach to observe all state changes using observeForever
- Now validates the complete flow: Loading → Success states
- Uses advanceUntilIdle() for proper async test handling

dcalhoun reviewed (approved):
This approach feels much better. Thank you for completing this refactor. 

These test plan succeeded for me. 🎉
